### ISSUE TITLE
Is it possible for the target to service MPI_Get's in parallel?

### ISSUE THREAD

**Issue by frizwi** [open-mpi/ompi#9508]
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
Using MPI v4.1.1 on Linux

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

Built from source tarball, no other configure flags other than **--prefix**

-----------------------------

## Details of the problem

Its more of a question: with point-to-point communication, either the non-blocking sends or blocking sends in a thread each can be used to parallelise transfers but is there any way to setup/build/configure the OpenMPI library to do this at the target?

I have a master process and many workers, each worker process performs and MPI_Get with the master process as the target via passive synchronisation. The issue is that as the number of workers increases so does the walltime for the transfers, so I'd like the target to service the Get's in parallel threads - is that possible? With point-to-point comms, I could either use non-blocking sends or blocking sends in a parallel OpenMP sections to achieve this but not clear if I can do this in any way via RMA? Is there some trick that can be done with active synchronisation? The only other thing I can think of is to do an MPI_Put from the master in parallel - will that work?

Thanks in advance for any suggestions, or please point me to some doc, if available. The MPI spec. doesn't really address my issue so I'm hoping to get an understanding of how OpenMPI actually implements RMA to be able to fine tune my application

**Comment by frizwi** [comment]
Using `MPI_Win_create` to create Windows, saves on modifying hundreds of arrays in existing code that uses malloc. Application needs to work in non MPI mode too but will add wrapper functions if it turns out to be a big performance gain

Also a follow up question, on the worker nodes I'm doing something like this call
`err = MPI_Win_create(lpatm, szS*sizeof(double), sizeof(double),
			 MPI_INFO_NULL, MPI_COMM_WORLD, &win);`
where `lpatm` is the receive buffer for each worker and used in the `MPI_Get` call. But strictly speaking, I don't need to expose the worker's memory so can just pass in `NULL`. However, is it doing any harm? I.e. any issue with it being copied into a public area in memory even though its never referenced by any other process?

**Comment by devreal** [comment]
Mhh, when breaking it down to (3 operations (lock, get, unlock) x 5 iterations x N worker) I end up with ~10ms per operation in both cases, which seems quite slow. I wanted to test your example both with and without xpmem but it doesn't seem to be self-contained so I fail to compile it, unfortunately. Have you had a chance to run with UCX and/or xpmem?

**Comment by devreal** [comment]
@frizwi I ran your example on our IB system and get the following runtimes:

2W:
```
001 time = 0.056773
002 time = 0.055926
```

4W:
```
001 time = 0.041799
002 time = 0.038467
003 time = 0.041928
004 time = 0.042871
```

8W:
```
Timing stats:
001 time = 0.024796
002 time = 0.025860
003 time = 0.026318
004 time = 0.025359
005 time = 0.024309
006 time = 0.025429
007 time = 0.024707
008 time = 0.023455
```

That seems to do be what you expect, right? Looking at your example, I would encourage you to move the shared locks out of the main loop and use `MPI_Rget`+`MPI_Wait` instead, as that saves a bit of overhead.

Bottom line: `MPI_Win_create` in combination with shared memory only works well under certain circumstances (e.g., when using xpmem) but it should work as expected on networks supporting RDMA, like IB. It's unfortunate that MPI currently won't tell you what it does under the hood and whether it can serve your gets efficiently or not...

**Comment by frizwi** [comment]
Thanks @devreal so it sounds like it *is possible* to get the scaling that I'm looking for, as per your post above. The bad news is that I'm unable to reproduce the scaling on my cluster. I've attached the output of ompi_info, if you could post yours, then I could try and get it built the same way.
[ompi-info.gadi.txt](https://github.com/open-mpi/ompi/files/7457666/ompi-info.gadi.txt)

Are you running it all on one node or distributed across the cluster. I was doing my experiments in an interactive job on the one node.

In any case, I've made the changes to using MPI_Rget+MPI_Wait, as per your suggestion like so (relevant section on the worker)
```
/* Assuming this lock is shared */
err = MPI_Win_lock_all(0, win);
MPI_ERR(err);
  
/* sleep to simulate loop and barrier */
for (s=0; s<NSTEPS; s++) {
   double t0, t1, dt;
      
    /* Sync with master */
    MPI_Barrier(MPI_COMM_WORLD);

    t0 = MPI_Wtime();

    err = MPI_Rget(lpatm, szS, MPI_DOUBLE, 0, 0, 1, dtype, win, &req);
    MPI_ERR(err);
      
    err = MPI_Wait(&req, MPI_STATUS_IGNORE);
    MPI_ERR(err);
      
    t1 = MPI_Wtime();
    dt = t1-t0;
    
    MPI_Gather(&dt, 1, MPI_DOUBLE, NULL, 1, MPI_DOUBLE, 0, MPI_COMM_WORLD);
 } // for NSTEPS
err = MPI_Win_unlock_all(win);
MPI_ERR(err);
```
Is this what you mean?

So the Window is only ever locked once for the duration of each of the worker processes, and then `Rget` + `Wait` take care of the transfers, protection of the target's memory is handled by some other means, in this case MPI_Barrier

It's evident from your example that there is some parallelism/shared memory transfer going on so that gives me hope that with the right settings/build it can be done!

Just to recap, is this the best approach for the master->worker data transfers w/MPI RMA? It certainly fits in with the workflow that I have in my application.

**Comment by frizwi** [comment]
On a different cluster, I get these timings:
2W:
```
001 time = 0.019260
002 time = 0.019432
```
4W:
```
001 time = 0.014843
002 time = 0.013651
003 time = 0.014758
004 time = 0.016316
```

8W:
```
001 time = 0.010753
002 time = 0.010745
003 time = 0.011163
004 time = 0.010900
005 time = 0.010377
006 time = 0.011214
007 time = 0.010494
008 time = 0.010068
```
Not as good a scaling as yours, but atleast it's not _descaling_  like in the other gadi cluster I reported about a week ago!

It's ompi_info is here: 
[ompi-info.pearcey.txt](https://github.com/open-mpi/ompi/files/7457730/ompi-info.pearcey.txt)

**Comment by devreal** [comment]
@frizwi Yes, the changes to the code are what I was thinking (although my understanding is that you're only ever reading from process with rank 0, so taking a shared shared for that process would be sufficient). 

I could recreate the scaling on both shared memory and with a single process per node.

As I said earlier, `MPI_Win_create` won't work well on shared memory unless you've built either Open MPI or UCX against xpmem [1]. Both my Open MPI and UCX are built with xpmem support. If the xpmem kernel module is not installed on your system you would have to convince your admin to install it. Only then can a process access another process' memory directly (using knem [2] might work as well but knem doesn't seem to be installed on our systems) On multiple nodes, the IB hardware will take care of accessing memory on a different node. In both cases, you will get the desired behavior of transfers without requiring help from the target process.

[1] https://github.com/hjelmn/xpmem
[2] https://knem.gitlabpages.inria.fr/doc/knem-api.html

**Comment by frizwi** [comment]
@devreal Yes, that's right, only rank 0 will ever write into the memory referenced by that win and all other ranks will only ever read. In ```MPI_Win_create``` rank > 0 pass NULL. ```MPI_Win_lock``` seems to work as it's not true lock in the mutex sense otherwise this wouldn't work. I'll only ever call it once at the start of the application for each worker rank.

Finally, any assert values that help here?

**Comment by devreal** [comment]
I cannot think of any info key that would apply here. 

I would appreciate if you could keep us updated on whether performance meets your expectations once you run your application on the target cluster. In the mean-time, feel free to close this issue if you feel like it's been solved.

**Comment by frizwi** [comment]
Thanks @devreal, will keep you posted.

Just one last thing before I close this ticket, I have uploaded higher number or worker windows here:
https://cloudstor.aarnet.edu.au/plus/s/TbRz4TcwEkKJNNc

So am curious how far we can take the scaling, 2-4-8 seems to scale nicely, as per your post earlier but how about 40-95 and up to 187? Could you try it on your cluster. Does ~100W get to Infiniband's theoretical limit?
